lib/datasets.go

```diff
@@ -203,22 +206,129 @@ func (m *DatasetMethods) ListRawRefs(p *ListParams, text *string) error {
 
 // GetParams defines parameters for looking up the head or body of a dataset
 type GetParams struct {
-	// Refstr to get, representing a dataset ref to be parsed
-	Refstr   string
-	Selector string
+	Refstr string `json:"ref"`
+
+	Selector string `json:"selector"`
 
 	// read from a filesystem link instead of stored version
-	Format       string
-	FormatConfig dataset.FormatConfig
+	Format       string               `json:"format"`
+	FormatConfig dataset.FormatConfig `json:"format_config"`
 
-	Limit, Offset int
-	All           bool
+	Limit  int  `json:"limit"`
+	Offset int  `json:"offset"`
+	All    bool `json:"all"`
 
 	// outfile is a filename to save the dataset to
-	Outfile string
+	Outfile string `json:"outfile"`
 	// whether to generate a filename from the dataset name instead
-	GenFilename bool
-	Remote      string
+	GenFilename bool   `json:"genfilename"`
+	Remote      string `json:"remote"`
+}
+
+// SetNonZeroDefaults assigns default values
+func (p *GetParams) SetNonZeroDefaults() {
+	if p.Format == "" {
+		p.Format = "json"
+	}
+}
+
+var validSelector = regexp.MustCompile(`^[\w-\.]*[\w]$`)
+
+func parseSelector(selector string) (string, string, error) {
+	if selector == "" {
+		return "", "", nil
+	}
+
+	format := ""
+
+	if strings.HasSuffix(selector, ".json") {
+		format = "json"
+	}
+	if strings.HasSuffix(selector, ".csv") {
+		format = "csv"
+	}
+	if strings.HasSuffix(selector, ".zip") {
+		format = "zip"
+	}
+
+	if format != "" {
+		selector = selector[:len(selector)-len(format)-1]
+	}
+
+	match := validSelector.FindString(selector)
+	if match == "" || len(match) != len(selector) {
+		return "", "", fmt.Errorf("could not parse request: invalid selector")
+	}
+	return selector, format, nil
+}
+
+func arrayContains(subject []string, target string) bool {
+	for _, v := range subject {
+		if v == target {
+			return true
+		}
+	}
+	return false
+}
+
+// UnmarshalFromRequest implements a custom deserialization-from-HTTP request
+func (p *GetParams) UnmarshalFromRequest(r *http.Request) error {
+	mvars := mux.Vars(r)
+
+	if p == nil {
+		p = &GetParams{}
+	}
+
+	params := *p
+	params.Refstr = r.FormValue("refstr")
+
+	ref, err := dsref.Parse(params.Refstr)
```

Comment:
I have some subtle improvements for this but will land them with the /save work. Mostly guards for preventing overwriting already set values from query params. Essentially body data > formvalue data. It's less important in this specific endpoint but is a good example as others are not as tolerant. Especially since the `stripServerSideQueryParams(r)` can be easily forgotten and add subtle bugs.

Reply:
very much agree the `stripServerSideQueryParams` is a hack that needs fixing / cleanup. IMHO if we stick with the pattern of overloading query parameters, we should consider modifying the `apiutil` package to auto-remove a set list of params before writing paginated responses.